### ISSUE TITLE
Perf: DateTime.Parse(string, CultureInfo) 90x slower on Linux

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#3811]
I can post or email the individual iteration results if desired; the values shown are the medians.

| test name                                                                      |  date ran  |  Windows time |  Linux time |  linux/windows | 
|--------------------------------------------------------------------------------|------------|---------------|-------------|----------------| 
| System.Globalization.Tests.Perf_DateTimeCultureInfo.Parse(culturestring: "ja") | 10/13/2015 | 3.926445575   | 269.9053    | 68.7403645     | 
| System.Globalization.Tests.Perf_DateTimeCultureInfo.Parse(culturestring: "")   | 10/13/2015 | 3.748826086   | 151.5521    | 40.42654861    | 
| System.Globalization.Tests.Perf_DateTimeCultureInfo.Parse(culturestring: "fr") | 10/13/2015 | 3.861441974   | 135.64      | 35.12677412    | 
| System.Globalization.Tests.Perf_DateTimeCultureInfo.Parse(culturestring: "da") | 10/13/2015 | 5.109054945   | 113.3454    | 22.18519887    | 

Perf test:
```
[Benchmark]
[InlineData("fr")]
[InlineData("da")]
[InlineData("ja")]
[InlineData("")]
public void Parse(string culturestring)
{
    CultureInfo cultureInfo = new CultureInfo(culturestring);
    foreach (var iteration in Benchmark.Iterations)
        using (iteration.StartMeasurement())
        {
            for (int i = 0; i < innerIterations; i++)
            {
                DateTime.Parse("10/10/2010 12:00:00 AM", cultureInfo);
                DateTime.Parse("10/10/2010 12:00:00 AM", cultureInfo);
                DateTime.Parse("10/10/2010 12:00:00 AM", cultureInfo);
                DateTime.Parse("10/10/2010 12:00:00 AM", cultureInfo);
            }
        }
}
```

**Comment by ianhays** [comment]
@eerhardt @ellismg

**Comment by stephentoub** [comment]
cc: @steveharter

**Comment by steveharter** [comment]
@ianhays is 90x on windows with coreclr, or desktop clr? In a similar test, I'm getting:
coreclr+windows 4.3x faster than coreclr+linux
desktopclr+windows 18x-25x faster than coreclr+linux
This includes using a simple cpu-only baseline looping test used to normalize the results due to different cpu speeds and\or differences in clr jitting\etc.

**Comment by steveharter** [comment]
@ellismg @stephentoub since you two have done some string optimizations in Linux\ICU.

The poor perf is due to hundreds of string compares that call to Native code. A single DateTime.Parse calls CompareInfo.CompareString(...) ~690 times. Hacking up a simpler string comparison, a loop of 400,000 DateTime.Parse calls goes from 20,692 to 386 milliseconds.

The main culprit for the high number of calls is DateTimeFormatInfo.InsertHash() encountering collisions in a custom hashtable.

Thus we can optimize the native string calls and\or improve the hashtable.

**Comment by stephentoub** [comment]
Thanks, @steveharter.  Are the number of managed to native calls and the number of hashtable collisions significantly different between Windows and Linux for some reason?  Or are those operations just significantly more expensive on Linux than on Windows?

**Comment by steveharter** [comment]
@stephentoub the number of string compare calls between Linux and Windows is the same. The raw string comparison is ~100x slower on Linux using this sample code:
cultureInfo.CompareInfo.Compare("ABC", "ABc", CompareOptions.IgnoreCase)

**Comment by stephentoub** [comment]
@steveharter, thanks, makes sense then.  Seems like we have two things to be addressed then:
- Investigating why DateTime.Parse is doing so many string comparisons and whether that's really necessary
- Fixing string comparison performance.  I believe @ellismg is already exploring changes here, e.g. caching the collators in the CompareInfos rather than creating them and destroying them each time, similar to how we do on Windows.

**Comment by steveharter** [comment]
For this issue, @stephentoub I'll focus on reducing the number of string comparisons. I'll let @ellismg explore improving the raw string comparison performance.

The strings being compared are days of the week, month names, etc. due to hashtable bucket collision. So we can reduce that by changing the hashtable implementation, and\or perhaps deferring creating that hashtable until actually necessary (since in the test code, day and month names weren't even used in the parse string).

**Comment by stephentoub** [comment]
Sounds good.  Thanks, Steve.

**Comment by steveharter** [comment]
Current progress reduces the number of string comparisons (for the specific tests) by 80%, effectively increasing performance by 77% for Linux (meaning tests run in ~23% of the time they used to). They are still significantly slower than Windows however.

Three optimizations:
1) Single character token comparisons check for Ascii and do direct check without calling lower-level culture comparisons. This is the main win for the tests here as it short-circuits the : and / time and date separators.
2) When creating the tokens in DateTimeFormatInfo, a 'return' statement was added which reduces the number of tokens by eliminating duplicates (for 'en-US', from 114 to 59, for 'fr' from 114 to 92). Not measured by the current tests, but affects the first run of .Parse and avoids some subsequent string comparisons.
3) When comparing a day or month being parsed like "June" against a token like "January", comparisons are avoided if the strings are not the same length (technically if June*** doesn't have a space after it). Not directly measured by the current tests.

**Comment by stephentoub** [comment]
Sounds like good progress, Steve, thanks.

**Comment by steveharter** [comment]
Fixed between #2112 and this issue which removed many of the string comparison calls.
